Automattic/pocket-casts-ios
# Update SwiftGen to always generate nested enums

**TL;DR: This diff is huge, but is all machine generated!** It's the result of a one line change in `swiftgen.yml`, `swiftgen` regenerating `Strings+Generated.swift` and me find and replacing usages of `L10n.<key>` into `L10n.Localizable.<key>`. If CI can build the app, then my changes are correct.

**Update** 🤔 I thought the intents strings were accessed somewhere in the codebase, but I might have been mistaken. #176  shows that the build works even though the existing code-generated values have been removed from `L10n`. 

I still think there's use in this PR just in case we'll have more `.strings` files in the future. But I can see how it might pull people off because of the extra typing (`L10n.Localizable` vs. the shorter `L10n`)... @Automattic/pocket-casts-ios keen to here from you folks whether you are okay with merging or would rather keep the `L10n` setup as it is right now.

## Description

Part of our localization tooling involves having multiple `.strings` files for the devs to update, and merging them all into a single `.strings` to upload to GlotPress for translation.

In the case of Pocket Casts iOS, these files would be `Localizable.strings` and `Intents.strings`. However, the current SwiftGen setup doesn't distinguish between the two, because 1) the keys [from `Intents.strings`](https://github.com/Automattic/pocket-casts-ios/blob/d5bb891345ca9d0e5d3800dcbdead5f9fc44d7f9/podcasts/en.lproj/Intents.strings) are [duplicated](https://github.com/Automattic/pocket-casts-ios/blob/d5bb891345ca9d0e5d3800dcbdead5f9fc44d7f9/podcasts/en.lproj/Localizable.strings#L2582-L2721) in `Localizable.strings` and 2) [SwiftGen is not configured to read from `Intents.strings`](https://github.com/Automattic/pocket-casts-ios/blob/d5bb891345ca9d0e5d3800dcbdead5f9fc44d7f9/swiftgen.yml#L13-L15).

For our localization automation to work we need to distinguish between `Localizable.strings` and `Intents.strings`. This PR is the first step in the process to get there.

When SwiftGen reads from multiple files, it generates nested enums for each. This PR configures SwiftGen to generated the nested enum even though it reads only one file. Once this lands on `trunk`, I'll followup with one that extracts `Intents.string`. 

To make the review easier, I recommend jumping directly to the diff for:

- [`swiftgen.yml`](https://github.com/Automattic/pocket-casts-ios/pull/174/files#diff-862e00ea4e50f62a96ff7240c16ac81be353c68c3ed94f08c1601154cc550e86) to see the new configuration parameter (`forceFileNameEnum: true`)
- [`podcast/Strings+Generated.swift`](https://github.com/Automattic/pocket-casts-ios/pull/174/files#diff-5a1765dc22ed147e4fd6ad7b269be5249cd7dfdbf00c80395540e006b9134a15) (_this might take a while for GitHub to show_) to check that nothing has changed other than for the new nested enum `Localizable`:

<img width="1159" alt="Screen Shot 2022-08-23 at 12 19 10 pm" src="https://user-images.githubusercontent.com/1218433/186054909-a3420ee8-5dbd-40d8-baf4-f4eb4775543e.png">


Also notice the changes in [`podcast/Strings+L10n.swift`](https://github.com/Automattic/pocket-casts-ios/pull/174/files#diff-2585793f7ae07878d4e5129bdde48a08538a40f547b5880c6efd0e3d4bb1de2f). This, as far as I can tell, is a custom addition on top of the SwiftGen-code-generated strings with utils and some strings that are yet to be translated.

## To test

I decided to split the PRs to keep the review sane. As you can see, the diff here is _huge_ because the SwiftGen configuration change regenerated most of `L10n` constants used for localization. **But no code logic has changed**, I merely renamed the `L10n.<key>` usages to `L10n.Localizable.<key>`. So it should be safe to trust the app works if CI because CI is green. If I had missed something or made a typo, the app wouldn't compile.

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE_NOTES.md` if necessary. – N.A.

## Next steps

- Remove the `Intents.strings` keys from `Localizable.strings` and configure SwiftGen to read that file, too
- Rebase #94 on top of these changes and make it freeze the strings in a dedicated file (otherwise we'll return to the situation of having duplicated keys in `Localizable.strings`)

Side note: @leandroalonso and I briefly chatted about the possibility of moving away from the SwiftGen setup and adopt the approach other A8c apps follow: Devs use `NSLocalizableString("reverse.DNS.key", value: "some string", comment: "...")` and our automation extracts all the keys into the correct `.strings` at code freeze time. The work that I'm doing will make this approach possible (see the part of #94 where the code mentions not calling `ios_generate_strings_file_from_code`) if you folks will wish to adopt the approach.
